### PR TITLE
Simple get

### PR DESCRIPTION
**Pull request by jchung722**

### CODE REVIEW

File: src/main/java/com/eighthlight/Server.java [hunk 1]
```diff
@@ -4,20 +4,41 @@
...
+    public void readRequest(){
+        try {
+            receivedData = serverSocket.receiveData();
+            receivedRequest = new Request(receivedData);
+        } catch(IOException e){
+            System.out.println(e.getMessage());
+        }
+    }
+
+    public void sendResponse(){
```
Comment (cmvandrevala) [inline src/main/java/com/eighthlight/Server.java]: Right now, your Request class exposes method, uri, and version through getter methods. Then, the Server class asks Request for that information and does a comparison with the incoming request. In a sense, the Server knows the internals of Request, which breaks encapsulation of information. Instead, consider deleting the getters on Request and replacing them with methods specific to the comparison that you are doing. For example.

```
class Request {
  public bool isAGetRequest()
   public bool isHttpVersionOnePointOne()
  // etc...
}
```

Comment (cmvandrevala) [reply src/main/java/com/eighthlight/Server.java]: This is often called "tell, don't ask". Don't ask a class for internal information and then do work on it. Instead, tell a class to do something and use the result (in this case, tell Request to make a comparison and use the boolean)

File: src/test/java/com/eighthlight/SocketWrapperSpy.java [hunk 2]
```diff
@@ -1,17 +1,49 @@
...
+    private BufferedReader in;
+    private String receivedData;
+    private String sentData;
+
+    public SocketWrapperSpy(BufferedReader in){
+        this.in = in;
+    }
+
+    public SocketWrapperSpy(){
+        this(new BufferedReader(new StringReader("")));
```
Comment (cmvandrevala) [inline src/test/java/com/eighthlight/SocketWrapperSpy.java]: What does wrapping BufferedReader in "this" do? Not saying it is right or wrong, I just don't remember what this syntax does

File: src/test/java/com/eighthlight/ServerTest.java [hunk 3]
```diff
@@ -2,15 +2,31 @@
...
 import org.junit.Test;
 import org.junit.Assert;
+import java.io.*;
 
 public class ServerTest {
+
+    @Test
+    public void receivesRequestMessageFromClient() {
+        BufferedReader in = new BufferedReader(new StringReader("GET /simple_get HTTP/1.1"));
+        SocketWrapperSpy testSocketWrapper = new SocketWrapperSpy(in);
```
Comment (cmvandrevala) [inline src/test/java/com/eighthlight/ServerTest.java]: Nice use of a spy here. Very neat

File: src/test/java/com/eighthlight/RequestTest.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
+package com.eighthlight;
+
+import org.junit.Test;
+import org.junit.Assert;
+
+public class RequestTest {
```
Comment (cmvandrevala) [inline src/test/java/com/eighthlight/RequestTest.java]: See my previous comment on encapsulation. These tests might change if you make a refactoring

File: src/test/java/com/eighthlight/ServerTest.java [hunk 5]
```diff
@@ -2,15 +2,31 @@
...
+        Server testServer = new Server(testSocketWrapper);
+
+        testServer.readRequest();
+
+        Assert.assertEquals("GET /simple_get HTTP/1.1", testSocketWrapper.getReceivedData());
```
Comment (bgerstle) [inline src/test/java/com/eighthlight/ServerTest.java]: I'm confused. Why are you asserting that your spy received the data you set up in the buffered reader?

File: src/main/java/com/eighthlight/Server.java (hunk 1 above)
Comment (bgerstle) [reply src/main/java/com/eighthlight/Server.java]: FWIW, I don't think this breaks encapsulation. I would have done a similar thing (treat `Request` as a "dumb" data object) and check its fields to decide what I should do. However, I would point out that the _responsibilities_ of differentiating (i.e. routing) requests and determining a response should live elsewhere so it can be tested in isolation.

File: src/test/java/com/eighthlight/SocketWrapperSpy.java (hunk 2 above)
Comment (jchung722) [reply src/test/java/com/eighthlight/SocketWrapperSpy.java]: Oh this is just a way for the constructor to work if there is no argument passed through; it passes in the buffered reader as if it is an argument. This was a way to compensate for the fact that java doesn't do default parameters.

Comment (bgerstle) [reply src/test/java/com/eighthlight/SocketWrapperSpy.java]: 👍 in Swift/Obj-C this is called the "designated initializer" pattern, where you have multiple "overloads" for your constructor/initializer that all call down to the same root initializer which accepts all the class's dependencies.

Comment (bgerstle) [reply src/test/java/com/eighthlight/SocketWrapperSpy.java]: @cmvandrevala this is effectively calling the constructor on line 12 which takes a `BufferedReader` argument

File: src/test/java/com/eighthlight/ServerTest.java (hunk 5 above)
Comment (jchung722) [reply src/test/java/com/eighthlight/ServerTest.java]: hmm well I figured that I would want to make sure data was being received from the client socket properly? Is this a bad test?

Comment (bgerstle) [reply src/test/java/com/eighthlight/ServerTest.java]: >  to make sure data was being received from the client socket properly? 

Received by whom, and what should they be doing with it?

What you've written so far only asserts that `receiveData()` was called on the spy, which results in the spy reading the data from `in`, and returning it from `getReceivedData`.

While it's common practice to assert spies received expected calls, the _main objective_ of your test should be verify behaviors of the _test subject_ (in this case, the server. This is different from verifying implementation details (i.e. "it called a method"). 

So, let me rephrase my initial question: what `Server` behavior does this test verify?

File: src/test/java/com/eighthlight/ServerTest.java [hunk 6]
```diff
@@ -2,15 +2,31 @@
...
-    public void createsSocketAndListensForClient() {
-        SocketWrapperSpy testSocketWrapper = new SocketWrapperSpy();
+    public void sendsOKResponseDataBackToClientForSimpleGet() {
+        BufferedReader in = new BufferedReader(new StringReader("GET /simple_get HTTP/1.1"));
+        SocketWrapperSpy testSocketWrapper = new SocketWrapperSpy(in);
         Server testServer = new Server(testSocketWrapper);
 
-        testServer.start(5000);
+        testServer.readRequest();
+        testServer.sendResponse();
```
Comment (bgerstle) [inline src/test/java/com/eighthlight/ServerTest.java]: Why do you need to call one after the other here? Is there a higher-level method for this read/write behavior?

File: src/test/java/com/eighthlight/ServerTest.java [hunk 7]
```diff
@@ -2,15 +2,31 @@
...
 
-        Assert.assertTrue(testSocketWrapper.wasCreatedAndListened());
+        Assert.assertEquals("HTTP/1.1 200 OK", testSocketWrapper.getSentData());
```
Comment (bgerstle) [inline src/test/java/com/eighthlight/ServerTest.java]: @jchung722 in contrast to the test we're discussing above, this is a great example of mocking done well. Here you're asserting a _behavior of the test subject_ (writing the expected response to its client's socket) by collecting the sent data using a spy. 💯

Comment (jchung722) [reply src/test/java/com/eighthlight/ServerTest.java]: I seee 👀

File: src/test/java/com/eighthlight/ServerTest.java (hunk 5 above)
Comment (jchung722) [reply src/test/java/com/eighthlight/ServerTest.java]: Hmm okay so would a better way to do this be to test the Server behavior of receiving a request? So it's not really necessary that we make sure got the data we did, but that we are properly receiving requests - so using the spy to verify that "receiveData" was called on the spy and data was received?

File: src/test/java/com/eighthlight/ServerTest.java (hunk 6 above)
Comment (jchung722) [reply src/test/java/com/eighthlight/ServerTest.java]: Yees the current higher level method is the start method, but I haven't figured out how to unit test a method with a while loop yet. Is this a call for another abstraction?

Comment (bgerstle) [reply src/test/java/com/eighthlight/ServerTest.java]: Yes, I would remove as much logic from the `while` loop as possible.

File: src/test/java/com/eighthlight/ServerTest.java (hunk 5 above)
Comment (bgerstle) [reply src/test/java/com/eighthlight/ServerTest.java]: I think "getting the data" is too low level. What is your `Server` _doing_ with the data?

Comment (bgerstle) [reply src/test/java/com/eighthlight/ServerTest.java]: I think part of what's making this hard to reason about is that `sendsOKResponseDataBackToClientForSimpleGet` isn't really a unit test. It's testing the integration of your logic that:

- Reads data from the socket and parses it into a request
- Decides how to handle that request
- Then creates, serializes, and writes the response to that request

How can you test these responsibilities in isolation?

Comment (bgerstle) [reply src/test/java/com/eighthlight/ServerTest.java]: And this test (`receivesRequestMessageFromClient`) should really be about the first responsibility: reading and parsing data into a `Request`.

Comment (jchung722) [reply src/test/java/com/eighthlight/ServerTest.java]: ookay so I think I just realized that my last "unit tests" were still integration tests. So here is a new push with unit tests there are less coupled to the sockets. Are these more on the right track?

File: src/main/java/com/eighthlight/Server.java [hunk 8]
```diff
@@ -4,41 +4,54 @@
...
                 serverSocket.createAndListen(port);
-                readRequest();
             }
         } catch(IOException e) {
             System.out.println(e.getMessage());
         }
     }
 
-    public void readRequest(){
+    public void handleRequest(){
```
Comment (bgerstle) [inline src/main/java/com/eighthlight/Server.java]: 👍

File: src/test/java/com/eighthlight/ServerTest.java [hunk 9]
```diff
@@ -7,14 +7,27 @@
...
     @Test
-    public void receivesRequestMessageFromClient() {
-        BufferedReader in = new BufferedReader(new StringReader("GET /simple_get HTTP/1.1"));
-        SocketWrapperSpy testSocketWrapper = new SocketWrapperSpy(in);
+    public void parsesDataFromClientIntoRequest() {
+        SocketWrapperSpy testSocketWrapper = new SocketWrapperSpy();
+        Server testServer = new Server(testSocketWrapper);
+        String requestMessage = "GET /simple_get HTTP/1.1";
+
+        Request request = testServer.parseRequestMessage(requestMessage);
```
Comment (bgerstle) [inline src/test/java/com/eighthlight/ServerTest.java]: This is on the right track, but is it really a `Server` test?

File: src/test/java/com/eighthlight/ServerTest.java [hunk 10]
```diff
@@ -7,14 +7,27 @@
...
+    }
+
+    @Test
+    public void createsOKResponseForSimpleGetRequest() {
+        SocketWrapperSpy testSocketWrapper = new SocketWrapperSpy();
         Server testServer = new Server(testSocketWrapper);
+        Request simpleGetRequest = new Request("GET /simple_get HTTP/1.1");
 
-        testServer.readRequest();
+        String response = testServer.createResponse(simpleGetRequest);
```
Comment (bgerstle) [inline src/test/java/com/eighthlight/ServerTest.java]: Also on the right track. I'm starting to see the beginnings of different responsibilities emerging here. I'd like you to have one more pass at refactoring the tests (and code) along the lines of these different responsibilities.

Doing this now will allow you to be much more precise when it comes time to add new features (and tests), so that you're touching the smallest amount of code possible for each change and each test.

File: src/test/java/com/eighthlight/SocketWrapperSpy.java (hunk 2 above)
Comment (cmvandrevala) [reply src/test/java/com/eighthlight/SocketWrapperSpy.java]: Ah, gotcha. Thanks all!
